Added the support for Native Intel SmithWaterman Aligner

pnvaidya opened the pull request:
Added support for the Native Intel SmithWaterman Aligner. Added the unitTest for the same. 

Closes #1812.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -10,14 +10,11 @@
 
 public final class SmithWatermanIntelAligner implements SmithWatermanAligner {
```
Needs class level javadoc explaining what this is, i.e. that it's an implementation of smith waterman aligner that's takes advantage of intel hardware optimizations.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -30,7 +27,7 @@
     /**
      * Create a new SW native pairwise aligner
```
could you add javadoc explaining when it will throw

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -0,0 +1,55 @@
+     */
+
+
+    private final SWAlignerNativeBinding SmithWaterman = new IntelSmithWaterman();
```
this variable should start with a lowercase S.  Only types and constants should start with a capital letter.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -0,0 +1,55 @@
+
+
+    /**
+     * return the stateless singleton instance of SmithWatermanIntelAligner
```
you removed the method but not the comment.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -0,0 +1,55 @@
+
+    /**
+     * Create a new SW native pairwise aligner
+     */
```
javadoc should include the fact that this can fail and explain why

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanJavaAligner.java:
```diff
@@ -378,4 +386,8 @@ private static CigarElement makeElement(final State state, final int length) {
 
+    @Override
+    public void close() {
+        logger.info(String.format("Total compute time in java Smith-Waterman : %.2f sec", totalComputeTime * 1e-9));
```
make this a debug level statement as well

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    @Override
+    public SmithWatermanAlignment align(final byte[] reference, final byte[] alternate, final SWParameters parameters, final SWOverhangStrategy overhangStrategy) {
+        return alignerWrapper.align(reference, alternate,parameters,overhangStrategy);
```
missing a few spaces between the parameters here

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanIntelAligner.java:
```diff
@@ -0,0 +1,55 @@
+import org.broadinstitute.hellbender.exceptions.UserException;
+
+
+public final class SmithWatermanIntelAligner implements SmithWatermanAligner {
```
this class needs a comment explaining what it is, i.e. it's a version of the SmithWatermanAligner that is optimized for intel hardware

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SWNativeAlignerWrapper.java:
```diff
@@ -1,30 +1,69 @@
+     */
     @Override
     public void close() {
+        logger.info(String.format("Total compute time in native Smith-Waterman : %.2f sec", totalComputeTime * 1e-9));
```
make this a debug level log message

lbergelson commented:
@pnvaidya Actually, one more request,  could go into the haplotype caller integration tests and specify to use `AVX_ENABLED` smith waterman?

pnvaidya commented:
I have corrected all the comments and modified code. For haplotypeCaller Tests should I add AVX_ENABLED for all tests?, apparently Travis is non AVX2 machine so it will skip the test if hardware is not available.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanAlignerAbstractUnitTest.java:
```diff
@@ -21,6 +21,8 @@
  */
+
+@Test
 public abstract class SmithWatermanAlignerAbstractUnitTest extends BaseTest {
```
After https://github.com/broadinstitute/gatk/pull/3475, this should extend `GATKBaseTest`

pnvaidya commented:
I have incorporated your changes, made default to not use FASTEST_AVAILABLE for smithwaterman and also updated the pull request to use latest GKL 0.8.1 which has the AVX512 support for smithwaterman as well. 
Thanks,

gspowley commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SWNativeAlignerWrapper.java:
```diff
@@ -1,30 +1,69 @@
+     */
     @Override
     public void close() {
+        logger.debug(String.format("Total compute time in native Smith-Waterman : %.2f sec", totalComputeTime * 1e-9));
```
@lbergelson, Can we use `info` for this message? Turning on `debug` messages creates huge log files and impacts performance. We have a similar info message for PairHMM performance, so it would be nice to have the same info for Smith-Waterman.

gspowley commented on src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanJavaAligner.java:
```diff
@@ -378,4 +386,8 @@ private static CigarElement makeElement(final State state, final int length) {
 
+    @Override
+    public void close() {
+        logger.debug(String.format("Total compute time in java Smith-Waterman : %.2f sec", totalComputeTime * 1e-9));
```
Same comment as above about using `info`.

droazen commented:
I ran with this branch over the weekend, and found that the GVCF output was IDENTICAL apart from the command line in the header. I think we can merge this!

droazen commented:
Fixed one typo in the log output -- merging.
